### PR TITLE
Detection Transformer Estimator

### PR DESCRIPTION
**Pull request by kieranfraser**
# Description

Adding **DE**tection **TR**ansformer (DETR) to ART to facilitate object detection using Transformers.

- DETR estimator added to art/estimators/object_detection
- requires additional method in art/estimators/pytorch.py which sets MultiheadAttention layer into train/eval mode
- associated tests added
- notebook demonstrating use of DETR estimator detecting objects and under adversarial patch attack added to notebooks/adversarial_patch/ and updated notebooks/ README

Paper: ["End-to-End Object Detection with Transformers"](https://arxiv.org/abs/2005.12872) 

Fixes # (issue)
#1638

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Test prediction of DETR
- [x] Test loss gradient
- [x] Test applying preprocessing defences
- [x] Test computation of losses (classification, bbox, giou)
- [x] Test DETR with PGD attack

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/estimators/object_detection/pytorch_detection_transformer.py [hunk 1]
```diff
@@ -0,0 +1,998 @@
...
+    import torch
+
+    from art.defences.postprocessor.postprocessor import Postprocessor
+    from art.defences.preprocessor.preprocessor import Preprocessor
+    from art.utils import CLIP_VALUES_TYPE, PREPROCESSING_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+def box_cxcywh_to_xyxy(x: "torch.Tensor"):
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_detection_transformer.py:42]: We need to mention the license of these functions in their docstrings.

Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_detection_transformer.py:42]: We need to mention the license of these functions in their docstrings.

File: art/estimators/object_detection/pytorch_detection_transformer.py [hunk 2]
```diff
@@ -0,0 +1,998 @@
...
+        :param postprocessing_defences: Postprocessing defence(s) to be applied by the estimator.
+        :param preprocessing: Tuple of the form `(subtrahend, divisor)` of floats or `np.ndarray` of values to be
+               used for data preprocessing. The first value will be subtracted from the input. The input will then
+               be divided by the second one.
+        :param device_type: Type of device to be used for model and tensors, if `cpu` run on CPU, if `gpu` run on GPU
+                            if available otherwise run on CPU.
+        """
+        import torch
+
+        class HungarianMatcher(torch.nn.Module):
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_detection_transformer.py:188]: We need to mention the license of these functions in their docstrings.

File: art/estimators/object_detection/pytorch_detection_transformer.py [hunk 3]
```diff
@@ -0,0 +1,998 @@
...
+                            "labels": torch.from_numpy(target["labels"]).type(torch.int64).to(self.device),
+                            "boxes": cxcy_norm.to(self.device),
+                            "scores": torch.from_numpy(target["scores"]).type(torch.float).to(self.device),
+                        }
+                    )
+
+        return x, targets
+
+
+class NestedTensor:
```
Comment (beat-buesser) [inline art/estimators/object_detection/pytorch_detection_transformer.py:917]: We need to mention the license of these functions in their docstrings.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @kieranfraser Thank you very much for your PR. I think we need to include the license of reused code somewhere. I think the patches don't have a lot of effects on the detections, is there a way how we could generate a stronger patch?

File: art/estimators/object_detection/pytorch_detection_transformer.py (hunk 1 above)
Comment (kieranfraser) [reply art/estimators/object_detection/pytorch_detection_transformer.py:42]: Added reference to Apache 2.0

File: art/estimators/object_detection/pytorch_detection_transformer.py (hunk 2 above)
Comment (kieranfraser) [reply art/estimators/object_detection/pytorch_detection_transformer.py:194]: Added reference to Apache 2.0

File: art/estimators/object_detection/pytorch_detection_transformer.py (hunk 3 above)
Comment (kieranfraser) [reply art/estimators/object_detection/pytorch_detection_transformer.py:925]: Added reference to Apache 2.0

**Comment by kieranfraser** [comment]
> Hi @kieranfraser Thank you very much for your PR. I think we need to include the license of reused code somewhere. I think the patches don't have a lot of effects on the detections, is there a way how we could generate a stronger patch?

Thanks @beat-buesser. I have added references to the license (Apache 2.0) in the `Docstrings` of reused code. Also updated the notebook with stronger Adversarial Patch attacks (targeted and untargeted). Seems that unless a fixed patch location is passed, the attack fails to converge to a patch that fools the DETR - plan to experiment further with this. But the notebook now demonstrates working patches as well as PGD for the DETR.
